ttl/ttl.go

```diff
@@ -13,29 +13,40 @@ import (
 
 var ErrNotFound = errors.New("not found")
 
+// value is an opaque structure that wraps a value that is store in the TTL
+// key/value map.
 type value struct {
 	value any
 
-	expired func(any, any) // called when expired
-	remove  func(any, any) // called when removed
+	expired func(any, any) // called when TTL expires
+	remove  func(any, any) // called when removed from map
+
+	timeoutExpired bool // set when this value has expired
 
 	// Value context
 	ctx    context.Context
 	cancel context.CancelFunc
 }
 
+// TTL is an opaque structure that wraps the TTL key/value map.
 type TTL struct {
 	mtx sync.Mutex
 
-	m map[any]value
+	autoDelete bool
+	m          map[any]*value
 }
 
-func New(capacity int) (*TTL, error) {
+// New creates a new TTL map with the provided capacity. If autoDelete is set
+// then the key will be deleted from the map on either the cancel or timeout
+// event.
+func New(capacity int, autoDelete bool) (*TTL, error) {
```

Comment:
Will more parameters like be added to this function in the future? If so, would it be worth using either an option pattern or creating a `ttl.Config` struct?

Reply:
IMO two is already too magical so I don't think/hope so.